### PR TITLE
Support on-demand incremental refresh

### PR DESCRIPTION
**Pull request by dai-chen**
### Description

Support on-demand incremental index refresh which is a new index refresh mode. Please check updated user manual for more details:

1. Added new section to explain 3 index refresh modes: https://github.com/dai-chen/opensearch-spark/blob/support-on-demand-incremental-refresh-rebased/docs/index.md#flint-index-refresh
2. Added new incremental refresh option: https://github.com/dai-chen/opensearch-spark/blob/support-on-demand-incremental-refresh-rebased/docs/index.md#create-index-options

#### Code Changes

**Most of the changed files are UT and IT. Reviewer can focus on the following changes**:

1. `FlintSparkIndexOptions`: added a new `incremental_refresh` option with default False value

2. `FlintSpark`: refactored complex `doRefresh()` method by the new `FlintSparkIndexRefresh` abstraction and its subclass `AutoIndexRefresh` and `FullIndexRefresh`

3. `IncrementalIndexRefresh`: added this subclass for new incremental mode

#### Class Diagram

FlintSparkIndexRefresh

#### Example

Create test table and skipping index with `incremental_refresh = true`:

```
spark-sql>
CREATE TABLE stream.refresh_test
(name STRING)
USING CSV
LOCATION 's3://refresh_test/';

INSERT INTO stream.refresh_test VALUES ('hello'), ('world');

CREATE INDEX test_idx ON stream.refresh_test (name)
WITH (
  incremental_refresh = true,
  checkpoint_location = 's3://checkpoints/refresh_test/'
);
```

At this moment, Flint index is empty with incremental refresh option stored in metadata:

<pre>
GET flint_myglue_stream_refresh_test_test_idx_index/_mapping
{
  "flint_myglue_stream_refresh_test_test_idx_index": {
    "mappings": {
      "_meta": {
        "latestId": "ZmxpbnRfbXlnbHVlX3N0cmVhbV9yZWZyZXNoX3Rlc3RfdGVzdF9pZHhfaW5kZXg=",
        "kind": "covering",
        "indexedColumns": [
          {
            "columnType": "string",
            "columnName": "name"
          }
        ],
        "name": "test_idx",
        "options": {
          "auto_refresh": "false",
          <b>"incremental_refresh": "true",</b>
          "checkpoint_location": "s3://checkpoints/refresh_test/"
        },
        "source": "myglue.stream.refresh_test",
        ...
</pre>

Trigger incremental refresh:

<pre>
<b># Similarly as full refresh, incremental refresh statement will exit once complete</b>
spark-sql> REFRESH INDEX test_idx ON stream.refresh_test;

POST flint_myglue_stream_refresh_test_test_idx_index/_search
    ...
    "hits": [
      {
        "_index": "flint_myglue_stream_refresh_test_test_idx_index",
        "_id": "1Ih0PY0BZG4KSy0Om1AJ",
        "_score": 1,
        "_source": {
          "name": "world"
        }
      },
      {
        "_index": "flint_myglue_stream_refresh_test_test_idx_index",
        "_id": "GTV0PY0BWdGHpYUCm64S",
        "_score": 1,
        "_source": {
          "name": "hello"
        }
      }
    ]

<b># Index state is back to active with timestamp updated</b>
POST .query_execution_request_myglue/_search
    ...
    "_source": {
          "version": "1.0",
          "latestId": "ZmxpbnRfbXlnbHVlX3N0cmVhbV9yZWZyZXNoX3Rlc3RfdGVzdF9pZHhfaW5kZXg=",
          "type": "flintindexstate",
          "state": "active",
          "applicationId": "unknown",
          "jobId": "unknown",
          "dataSourceName": "myglue",
          "jobStartTime": 1706133480199,
          "lastUpdateTime": 1706133493409,
          "error": ""
        }
</pre>

Insert more source data and trigger refresh again:

<pre>
spark-sql> INSERT INTO stream.refresh_test VALUES ('test');

spark-sql> REFRESH INDEX test_idx ON stream.refresh_test;

<b># Only latest new data is refreshed without duplicates</b>
POST flint_myglue_stream_refresh_test_test_idx_index/_search
    ...
    "hits": [
      {
        "_index": "flint_myglue_stream_refresh_test_test_idx_index",
        "_id": "1Yh7PY0BZG4KSy0OFFAw",
        "_score": 1,
        "_source": {
          <b>"name": "test"</b>
        }
      },
      {
        "_index": "flint_myglue_stream_refresh_test_test_idx_index",
        "_id": "1Ih0PY0BZG4KSy0Om1AJ",
        "_score": 1,
        "_source": {
          "name": "world"
        }
      },
      {
        "_index": "flint_myglue_stream_refresh_test_test_idx_index",
        "_id": "GTV0PY0BWdGHpYUCm64S",
        "_score": 1,
        "_source": {
          "name": "hello"
        }
      }
    ]
</pre>

### Issues Resolved

https://github.com/opensearch-project/opensearch-spark/issues/195

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: docs/index.md [hunk 1]
```diff
@@ -260,9 +271,10 @@ VACUUM MATERIALIZED VIEW alb_logs_metrics
...
 
-+ `auto_refresh`: triggers Incremental Refresh immediately after index create complete if true. Otherwise, user has to trigger Full Refresh by `REFRESH` statement manually.
++ `auto_refresh`: automatically refresh the index if set to true. Otherwise, user has to trigger refresh by `REFRESH` statement manually.
 + `refresh_interval`: a string as the time interval for incremental refresh, e.g. 1 minute, 10 seconds. This is only applicable when auto refresh enabled. Please check `org.apache.spark.unsafe.types.CalendarInterval` for valid duration identifiers. By default, next micro batch will be generated as soon as the previous one complete processing.
-+ `checkpoint_location`: a string as the location path for incremental refresh job checkpoint. The location has to be a path in an HDFS compatible file system and only applicable when auto refresh enabled. If unspecified, temporary checkpoint directory will be used and may result in checkpoint data lost upon restart.
++ `incremental_refresh`: incrementally refresh the index if set to true. Otherwise, fully refresh the entire index. This only applicable when auto refresh disabled.
```
Comment (penghuo) [inline docs/index.md:276]: what is default value incremental_refresh?

File: docs/index.md [hunk 2]
```diff
@@ -260,9 +271,10 @@ VACUUM MATERIALIZED VIEW alb_logs_metrics
 
 User can provide the following options in `WITH` clause of create statement:
 
-+ `auto_refresh`: triggers Incremental Refresh immediately after index create complete if true. Otherwise, user has to trigger Full Refresh by `REFRESH` statement manually.
++ `auto_refresh`: automatically refresh the index if set to true. Otherwise, user has to trigger refresh by `REFRESH` statement manually.
```
Comment (penghuo) [inline docs/index.md:274]: what is default value auto_refresh?

File: flint-spark-integration/src/main/scala/org/opensearch/flint/spark/refresh/AutoIndexRefresh.scala [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+    def addRefreshInterval(refreshInterval: Option[String]): DataStreamWriter[Row] = {
+      refreshInterval
+        .map(interval => dataStream.trigger(Trigger.ProcessingTime(interval)))
+        .getOrElse(dataStream)
+    }
+
+    def addAvailableNowTrigger(incrementalRefresh: Boolean): DataStreamWriter[Row] = {
+      if (incrementalRefresh) {
+        dataStream.trigger(Trigger.AvailableNow())
```
Comment (penghuo) [inline flint-spark-integration/src/main/scala/org/opensearch/flint/spark/refresh/AutoIndexRefresh.scala:101]: if no checkpoint location, incremental_refresh behavior is same as full_refresh?

File: docs/index.md (hunk 1 above)
Comment (dai-chen) [reply docs/index.md]: missed it and added it in https://github.com/opensearch-project/opensearch-spark/pull/234/commits/694d603a64727d570ec34e28d2b2da6d564c27e4

File: docs/index.md (hunk 2 above)
Comment (dai-chen) [reply docs/index.md]: missed it and added it in https://github.com/opensearch-project/opensearch-spark/pull/234/commits/694d603a64727d570ec34e28d2b2da6d564c27e4

File: flint-spark-integration/src/main/scala/org/opensearch/flint/spark/refresh/AutoIndexRefresh.scala (hunk 3 above)
Comment (dai-chen) [reply flint-spark-integration/src/main/scala/org/opensearch/flint/spark/refresh/AutoIndexRefresh.scala:101]: Added a quick validation in https://github.com/opensearch-project/opensearch-spark/pull/234/commits/694d603a64727d570ec34e28d2b2da6d564c27e4. will improve in https://github.com/opensearch-project/opensearch-spark/issues/65. Thanks!

**Pull request merged**